LEAN-2357

mbetamony opened the pull request:
for this ticket, i did the following: 
1- wrap sections into the 3 new section containers when moving sections to body
2- create a `section_container` node to handle nested sections behaviors
Here are some test results I did using `pressroom` (`importJats`/`exportJats`) found in testResults.zip

1. mnTest.zip is the input i used for `importJats`
2. manuscript-2.manuproj is the output of `importJats`, also the input of `exportJats`. Please note the `abstracts`, `body` and `backmatter` sections with their `elementIDs`
3. exportJatsOutput.xml is the output of `exportJats`. Note how the sections are surrounded between the new sections tags
(search for 
`abstracts` or `body` or `backmatter`)

[testResults.zip](https://github.com/Atypon-OpenSource/manuscripts-transform/files/11415609/testResults.zip)

Please let me know if we want to move on with this implementation or if you require any changes, so I can proceed with adding the necessary tests

Note: 
I realize the branch name is wrong, as I had some issue with GitHub. I'll fix that ASAP 
(LEAN is spelled LEA)

asouqi commented on src/jats/importer/jats-body-transformations.ts:
```diff
@@ -194,19 +207,34 @@ export const jatsBodyTransformations = {
     references: BibliographyItem[] | null,
     createElement: (tagName: string) => HTMLElement
   ) {
+    const bodyContainer = createSectionContainer('body', createElement)
+    const abstractsContainer = createSectionContainer(
+      'abstracts',
+      createElement
+    )
+    const backmatterContainer = createSectionContainer(
+      'backmatter',
+      createElement
+    )
+
+    const bodySections = doc.querySelectorAll('body > sec')
+    for (const section of bodySections) {
+      removeNodeFromParent(section)
+      bodyContainer.append(section)
+    }
```
I would prefer to separate this logic of wrapping all sections with our own special section to a new method. and we will call it after `moveSectionsToBody`

asouqi reviewed (changes requested):
@mbetamony could you please run `yarn test -u` to update snapshots, and you will need to update this [sample](https://github.com/Atypon-OpenSource/manuscripts-transform/blob/86bd2b5e77d0a6bdd36d640b89531799d4bab1b9/src/jats/__tests__/jats-roundtrip.test.ts#L166) for round-trips test case

mbetamony commented:
@mnatsheh @asouqi 
I published the following changes to the PR: 

1. removed `section_container` node and used the existing `section` node
2. fixed `exportJats` 
4. `section` node now  accepts `bibliography_section` inside its content ( ref-list wasn't showing when  exporting **JATS**)

Please let me know if you have any questions, I'll fix the tests once this approach is agreed on
Thank you

asouqi commented:
> section node now accepts bibliography_section inside its content ( ref-list wasn't showing when exporting JATS)

also we need to add `footnotes_section`   `graphical_abstract_section`

> I'll fix the tests once this approach is agreed on

for me, I'm ok with that

mbetamony commented:
I fixed the graphical-abstract section
for the footnotes, we're adding  them to the body section before  `moveSectionsToBody` in the `jats-body-transformations.ts`, so I fixed that. I didn't have to modify sections node for this
I'll test on more files and see if we need to change anything

mbetamony commented:
refactored some of the code, I didn't like the way I fixed fn sections before. 
I think the current code should work with everything, please let me know if you have any files/tests you'd like me to run 
I'll also fix the current test files, and fix merge conflicts

mbetamony commented:
There is a bug in graphical abstract sections, i'll  fix it ASAP.

mbetamony commented:
- Fixed all bugs I was able to find through testing
- Updated snapshots

mnatsheh commented on src/schema/nodes/section.ts:
```diff
@@ -58,7 +58,8 @@ const choosePageBreakStyle = (element: HTMLElement): number => {
   // NOTE: the schema needs paragraphs to be the default type, so they must explicitly come first
-  content: 'section_label? section_title (paragraph | element)* section*',
+  content:
+    'section_label? section_title (paragraph | element)* (graphical_abstract_section | section | footnotes_section | bibliography_section)*',
```
Instead of listing the different section types, can we use a group instead? if all these nodes have a group "sections", then we should be able to use "sections*".

mbetamony replied:
Done

mnatsheh commented on src/transformer/section-category.ts:
```diff
@@ -168,7 +174,7 @@ export const buildSectionCategory = (
 
     default:
-      return node.attrs.category || undefined
+      return node.attrs.category || node.attrs.group || undefined
```
Why is this needed?

mbetamony replied:
This was left unintentionally when removing the section_container node.

mnatsheh commented on src/jats/importer/jats-body-transformations.ts:
```diff
@@ -24,6 +24,23 @@ const removeNodeFromParent = (node: Element) =>
+  )
+
+  const title = createElement('title')
+  title.textContent = category ? chooseSecType(category) : ''
```
Can we use the section category title here?

mbetamony replied:
A couple of questions on this:
Do you want me to bring them from manuscripts-data or or use a switch statement similar to [chooseSectionCategoryFromTitle](https://github.com/Atypon-OpenSource/manuscripts-transform/blob/1556eb5b3613409da630626c3ed12de60c54ef0f/src/transformer/section-category.ts#L306)  or something else?
In manuscripts-data, the titles have been left empty for the new sections. I think its better If I add the following titles:
- backmatter -> backmatter
- body -> body
- abstracts-> abstracts

I'm not sure about capitalizing the first letter of the title, there are some inconsistencies in the data
Should I proceed with this change ?

mnatsheh replied:
You should use whatever is in manuscripts-data. It's fine if it's empty for now (I'm assuming that it will not break with an empty title)

mnatsheh commented on src/jats/importer/jats-body-transformations.ts:
```diff
@@ -194,19 +211,40 @@ export const jatsBodyTransformations = {
+      createElement
+    )
+    const bodySections = doc.querySelectorAll(
+      'body > sec:not([sec-type="backmatter"]), body > sec:not([sec-type])'
```
Is `sec[sec-type="backmatter"]` something that actually exist in the content?

mbetamony replied:
Yes. 
Before `moveSectionsToBody` we're calling `mapFootnotesToSections`. Which appends some of the `backmatter` sections to the body. So I created the `backmatter` container there so I don't have to query the sections again when moving sections to body.

mnatsheh replied:
Can we call `mapFootnotesToSections` inside `moveSectionsToBody`? I don't see a reason why it's separate. This was we can create all three containers in the same place.

mbetamony replied:
Sure, I also did refactored `moveSectionsToBody` into multiple functions for better readability, let me know if you prefer the old code.

mnatsheh commented on src/jats/jats-exporter.ts:
```diff
@@ -2025,29 +2024,52 @@ export class JATSExporter {
     }
   }
-
+  private moveBody = (body: HTMLElement) => {
```
Let's call this `unwrapBody`.

mnatsheh commented on src/jats/jats-exporter.ts:
```diff
@@ -2025,29 +2024,52 @@ export class JATSExporter {
+    if (!container) {
+      return
+    }
+    const sections = body.querySelectorAll(
```
You should be able to query `container` directly: `container.querySelectorAll('> sec')`.

mnatsheh commented on src/jats/jats-exporter.ts:
```diff
@@ -2025,29 +2024,52 @@ export class JATSExporter {
+    body.removeChild(container)
+  }
+  private removeBackContainer = (body: HTMLElement) => {
+    const container = body.querySelector(':scope > sec[sec-type="backmatter"]')
```
So we're assuming everything has been moved out of this section. Let's check if it is actually empty and log a warning if not.

mnatsheh replied:
Yeah that's fine.

mnatsheh commented on src/jats/jats-exporter.ts:
```diff
@@ -2025,29 +2024,52 @@ export class JATSExporter {
-        return true
-      }
+    const container = body.querySelector(':scope > sec[sec-type="abstracts"]')
+    const sections = container
+      ? body.querySelectorAll(':scope > sec[sec-type="abstracts"] > sec')
+      : body.querySelectorAll(':scope > sec')
+
+    const abstractSections = container
```
I prefer we combine those into a single if statement:

```
let abstractSections
if (container) {
    abstractSections = ...
} else {
    abstractSections = ...
}
```

mnatsheh commented on src/transformer/encode.ts:
```diff
@@ -671,7 +692,7 @@ const encoders: NodeEncoderMap = {
     path: path.concat([node.attrs.id]),
-    elementIDs: childElements(node)
+    elementIDs: sectionChildElements(node)
       .map((childNode) => childNode.attrs.id)
       .filter((id) => id),
```
Can we move these into the method? We can rename it to `sectionChildElementIds`.

mnatsheh commented on src/transformer/encode.ts:
```diff
@@ -253,6 +256,24 @@ const childElements = (node: ManuscriptNode): ManuscriptNode[] => {
+  node.forEach((childNode) => {
+    if (
+      !(
+        isSectionNode(childNode) ||
+        isBibliographySectionNode(childNode) ||
+        isGraphicalAbstractSectionNode(childNode) ||
+        isFootnotesSectionNode(childNode)
```
Can we check the group of the section category instead of checking the types one by one?

mbetamony replied:
Done

asouqi commented:
> Updated snapshots, but what updates do you mean? moving sections to their correct place?

no need as all test pass

mnatsheh commented on src/jats/jats-exporter.ts:
```diff
@@ -2025,30 +2024,57 @@ export class JATSExporter {
+    let abstractSections
+    if (container) {
+      abstractSections = Array.from(
+        body.querySelectorAll(':scope > sec[sec-type="abstracts"] > sec')
```
Let's do this the same way we did it for body (query the container).

mbetamony commented:
Sorry for the repeated commits, not sure what happened when I pushed this morning 
Do you wish I get rid of them?

asouqi commented on package.json:
```diff
@@ -29,7 +29,7 @@
   },
   "dependencies": {
-    "@manuscripts/json-schema": "^2.0.0",
+    "@manuscripts/json-schema": "^2.0.1",
```
use `2.0.1-LEAN-2416`

mbetamony replied:
Done
